### PR TITLE
feat(schemas): add async header

### LINKED ISSUE #347: Add async header for subscriber's questionnaire pages
Add async header for subscriber's questionnaire pages, i.e. header shows a questionnaire's name and a dropdown with the delete action.

image

### PR DESCRIPTION
**Pull request by zouxuoz**
resolve #347

### CODE REVIEW

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaDropdown.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    id: 'schema-header__destroy',
+    defaultMessage: 'Удалить анкету',
+  },
+
+  title: {
+    id: 'delete-modal__confirm',
+    defaultMessage: 'Подтвердите действие',
+  },
+
+  text: {
```
Comment (radiansz) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaDropdown.js]: Rename accordingly to component name

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    actions={[
+      <SchemaDropdownContainer schemaId={schemaId} />
+    ]} />
+);
+
+SchemaHeader.propTypes = {
+  intl: intlShape.isRequired,
+  schemaId: PropTypes.string.isRequired,
+  schema: PropTypes.shape({
+    data: PropTypes.shape({
```
Comment (radiansz) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js]: It' the mistake

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaLayout.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import SchemaHeader from './SchemaHeader';
+
+const SchemaLayout = ({ intl, schemaId, schema, children }) => (
+  <div>
+    <SchemaHeader intl={intl} schemaId={schemaId} schema={schema} />
+    {children}
+  </div>
+);
+
+SchemaLayout.propTypes = {
```
Comment (radiansz) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaLayout.js]: Lint problems.

File: src/routes/subscribers/routes/schemas/routes/show/containers/SchemaContainer.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+class SchemaContainer extends Component {
+  static propTypes = {
+    intl: intlShape.isRequired,
+    children: PropTypes.node.isRequired,
+    schemaId: PropTypes.string.isRequired,
+    fetching: PropTypes.bool.isRequired,
+    fetched: PropTypes.bool.isRequired,
+    schema: PropTypes.shape({
+      data: PropTypes.shape({
```
Comment (radiansz) [inline src/routes/subscribers/routes/schemas/routes/show/containers/SchemaContainer.js]: Why data is here?

File: src/routes/subscribers/routes/schemas/routes/show/index.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}));
+
+const SchemaRoutes = ({ pathname, pattern, params: { id } }) => (
+  <SchemaContainer schemaId={id}>
+    <Match exactly pattern={`${pattern}`} render={() => (
+      <Redirect to={`${pathname}/edit`} />
+    )} />
+
+    <Match pattern={`${pattern}/edit`} component={EditSchema} />
+    <DialogMatch id={'schema__destroy'} component={SchemaDestroyDialogContainer} />
```
Comment (radiansz) [inline src/routes/subscribers/routes/schemas/routes/show/index.js]: Why curly parentheses with id?

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaLayout.js (hunk 3 above)
Comment (zouxuoz) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaLayout.js]: fixed

File: src/routes/subscribers/routes/schemas/routes/show/index.js (hunk 5 above)
Comment (zouxuoz) [inline src/routes/subscribers/routes/schemas/routes/show/index.js]: fixed

File: src/routes/subscribers/routes/schemas/modules/schemas/__tests__/reducers.test.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  });
+
+  test('handles FETCH_SCHEMA_REQUEST', () => {
+    const initialState = { ...defaultState };
+
+    const expectedState = {
+      ...initialState,
+
+      byId: {
+        [schemaId]: {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/__tests__/reducers.test.js]: Can we collapse it in one line.

File: src/routes/subscribers/routes/schemas/modules/schemas/__tests__/reducers.test.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      },
+    };
+
+    const initialState = { ...defaultState, currentId: schemaId };
+
+    const expectedState = {
+      ...initialState,
+
+      byId: {
+        [payload.id]: {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/__tests__/reducers.test.js]: Can we collapse it in one line.

File: src/routes/subscribers/routes/schemas/modules/schemas/sagas.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  }
+}
+
+function* destroy() {
+  while (true) {
+    const { payload, meta: { resolve, reject, redirect } } = yield take(DESTROY_SCHEMA_REQUEST);
+
+    try {
+      yield call(api.schemas.destroy, payload);
+
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/sagas.js]: Remove the new line

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaDropdown.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  <div>
+    <Dropdown invert pull icon={<MoreHorizontalIcon />}>
+      <DropdownItem label={intl.formatMessage(messages.destroy)} onClick={openDestroyModal} />
+    </Dropdown>
+  </div>
+);
+
+SchemaDropdown.propTypes = {
+  intl: intlShape.isRequired,
+  openDestroyModal: PropTypes.func.isRequired,
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaDropdown.js]: openDistroyDialog

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+const messages = defineMessages({
+  backlink: {
+    id: 'schema-header__backlink',
+    defaultMessage: 'Назад к списку анкет',
+  },
+});
+
+const SchemaHeader = ({ intl, schemaId, schema: { name } }) => (
+  <SectionHeader
+    title={name}
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js]: Add a new line.

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    backlink={{
+      label: intl.formatMessage(messages.backlink),
+      to: '/subscribers/schemas',
+    }}
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js]: Add a new line.

File: src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    actions={[
+      <SchemaDropdownContainer schemaId={schemaId} />,
+    ]} />
+);
+
+SchemaHeader.propTypes = {
+  intl: intlShape.isRequired,
+  schemaId: PropTypes.string.isRequired,
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/components/SchemaHeader.js]: Add a new line.

File: src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDestroyDialogContainer.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    schemaId: PropTypes.string,
+    isOpen: PropTypes.bool.isRequired,
+    onClose: PropTypes.func.isRequired,
+  }
+
+  static contextTypes = {
+    router: PropTypes.object.isRequired,
+  }
+
+  onSubmit = () => {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDestroyDialogContainer.js]: Let's refactor it. Create two methods: `onSubmit` and `redirectAfterDestroy`

File: src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDropdownContainer.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+  static contextTypes = {
+    router: PropTypes.object.isRequired,
+  }
+
+  openDestroyModal = () => {
+    const { openDialog: open, schemaId } = this.props;
+
+    open({
+      id: 'schema__destroy',
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDropdownContainer.js]: add a new line

File: src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDropdownContainer.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+      params: {
+        schemaId,
+      },
+    });
+  }
+
+  render() {
+    const { intl } = this.props;
+
+    return (<SchemaDropdown
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDropdownContainer.js]: I think one line will work here.

File: src/routes/subscribers/routes/schemas/routes/show/index.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    )} />
+
+    <Match pattern={`${pattern}/edit`} component={EditSchema} />
+    <DialogMatch id="schema__destroy" component={SchemaDestroyDialogContainer} />
+  </SchemaContainer>
+);
+
+SchemaRoutes.propTypes = {
+  pathname: PropTypes.string.isRequired,
+  pattern: PropTypes.string.isRequired,
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/index.js]: add a new line.

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+  FETCH_SCHEMA_FAILURE,
+  DESTROY_SCHEMA_SUCCESS,
+} from './actions';
+
+const detailsById = (state = {}, { type, payload }) => {
+  switch (type) {
+
+    case FETCH_SCHEMA_REQUEST:
+      return {
+        ...state,
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Add a new line

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        [payload]: {
+          fetching: true,
+          fetched: false,
+          data: {},
+        },
+      };
+
+    case FETCH_SCHEMA_SUCCESS:
+      return {
+        ...state,
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Add a new line

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+  DESTROY_SCHEMA_SUCCESS,
+} from './actions';
+
+const detailsById = (state = {}, { type, payload }) => {
+  switch (type) {
+
+    case FETCH_SCHEMA_REQUEST:
+      return {
+        ...state,
+        [payload]: {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Single line is OK here

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+          fetching: true,
+          fetched: false,
+          data: {},
+        },
+      };
+
+    case FETCH_SCHEMA_SUCCESS:
+      return {
+        ...state,
+        [payload.id]: {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Single line is OK here

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+          fetching: false,
+          fetched: true,
+          data: payload.data,
+        },
+      };
+
+    case FETCH_SCHEMA_FAILURE:
+      return {
+        ...state,
+        [payload]: {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Single line is OK here

File: src/routes/subscribers/routes/schemas/routes/show/containers/SchemaContainer.js [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+  }
+}
+
+const mapDispatchToProps = {
+  pickSchemaAsCurrent,
+  fetchSchema,
+};
+
+const mapStateToProps = (state) => {
+  const currentSchema = state.schemas.detailsById[state.schemas.currentId] || {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/containers/SchemaContainer.js]: It's a reducer logic.

File: src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDestroyDialogContainer.js [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+    const { schemaId, destroySchema: destroy } = this.props;
+
+    return new Promise((resolve, reject) => {
+      destroy(schemaId, {
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/routes/show/containers/SchemaDestroyDialogContainer.js]: Please one line :-)

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+  DISMISS_CURRENT_SCHEMA,
+  FETCH_SCHEMA_REQUEST,
+  FETCH_SCHEMA_SUCCESS,
+  FETCH_SCHEMA_FAILURE,
+  DESTROY_SCHEMA_SUCCESS,
+} from './actions';
+
+const detailsById = (state = {}, { type, payload }) => {
+  switch (type) {
+
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Remove line

File: src/routes/subscribers/routes/schemas/modules/schemas/reducers.js [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+    case FETCH_SCHEMA_REQUEST:
```
Comment (rosendi) [inline src/routes/subscribers/routes/schemas/modules/schemas/reducers.js]: Add brackets

**Comment by radiansz** [comment]
lgtm

**Comment by rosendi** [comment]
super! lgtm!

**Comment by rosendi** [comment]
super! lgtm!

**Pull request merged**